src/adb_command.rs

```diff
@@ -0,0 +1,57 @@
+use std::process::{Command, Stdio};
+
+use failure::{err_msg, Error};
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct AdbCommand {
+    command: String,
+    device_id: Option<String>,
+    args: Vec<String>,
+}
+
+impl AdbCommand {
+    pub fn command(command: String) -> Self {
+        AdbCommand {
+            command,
+            device_id: None,
+            args: Vec::new(),
+        }
+    }
+
+    pub fn with_args(self, args: Vec<String>) -> Self {
+        AdbCommand { args, ..self }
+    }
+
+    pub fn with_device_id(self, device_id: Option<String>) -> Self {
+        AdbCommand { device_id, ..self }
+    }
+
+    pub fn execute(self) -> Result<String, Error> {
+        let mut command = Command::new("adb");
+
+        if let Some(device_id) = self.device_id {
```

Comment:
is the command still safe to execute if there is no device id given? will the command be executed for every attached device?
i have seen the test you are doing when calling into this functionality from `lib.rs`, but if we decide to expose this options publicly and for our own sake we should probably stop executing here IF adb will execute the command for every attached device to avoid data loss / other potential problems

Reply:
`adb` will execute the command without providing a `--device` only if exactly one device is connected to the system. Otherwise you have to specify which target device should be used.

Reply:
do we even need our own check for this if adb correctly handles the issue? or is the adb error not user friendly?

Reply:
e.g. `adb: unable to connect for backup: more than one device/emulator` the message is not too bad it is clear what the problem is but the user will don't know how to solve...

Reply:
not expensive, but less code = simpler project which is always good :) but I agree, we should keep it

Reply:
you're right, your message is better!
what do you think about calling adb and if the error you posted above is returned (I assume it has a specific error code) we are printing your error msg instead (i.e. moving it from a `custom check before the adb call` to a `handling of the adb call and printing a custom message`)

Reply:
If something went wronge (too many devices connected, no device connected, wrong parameters...), I'm always getting error-code 1 😞. In the error case I'm not getting the message from the stdout or stderr (don't know why atm) and parsing the message (if we could access it anyway) is pretty ugly...so in my point of view it is better to check that by hand.

Reply:
meh @ the error codes from adb. maybe we should switch to [fb-adb](https://github.com/facebook/fb-adb) at some point 😆

Reply:
Did some implementations with custom errors...but it is really not good do devicde the error type by string
```rust
#[derive(Fail, Debug)]
#[fail(display = "No connected device found.")]
pub struct NoConnectedDevice;

#[derive(Fail, Debug)]
#[fail(display = "More than one connected devices found.")]
pub struct MoreThanOneConnectedDevices;
```

``` rust
fn check_result(result: Result<(), Error>) -> Result<(), Error> {
    if let Err(e) = result {
        let cause = e.cause();

        if let Some(err) = cause.downcast_ref::<adb_command::MoreThanOneConnectedDevices>() {
            let error =
                "More than one device connected and no device provided.\n \
         Please execute adbackup again, with `--device` and one of the following device ids:\n";

            let error_message = format!("{}\n{}", error, get_printable_device_list()?);
            info!("{}", error_message);
            return Err(err_msg(error_message));
     ...
}
```

Just jumped over the readme of fb-adb...looks good, will create an issue for that to have a more detailed look but with a milestone 1.0.x